src/hivpy/hiv_testing.py

```diff
@@ -50,81 +55,125 @@ def __init__(self, **kwargs):
+        with specific symptoms and returns it. Presence of
+        an AIDS defining condition (ADC; any WHO4) in the previous time step,
+        tuberculosis (TB) in the previous two time steps,
+        and a non-TB WHO3 disease in the previous time step
+        all affect groupings and test probability.
         """
-        # set last test date
-        pop.set_present_variable(col.LAST_TEST_DATE, pop.date,
-                                 sub_pop=pop.apply_bool_mask(tested, sub_pop))
-        # "reset" dummy partner columns
-        pop.set_present_variable(col.NSTP_LAST_TEST, 0,
-                                 sub_pop=pop.apply_bool_mask(tested, sub_pop))
-        pop.set_present_variable(col.NP_LAST_TEST, 0,
-                                 sub_pop=pop.apply_bool_mask(tested, sub_pop))
+        # assume asymptomatic by default
+        prob_test = 0
+        # presence of ADC last time step
+        if adc_tm1:
+            prob_test = self.prob_test_who4
+        # presence of TB last time step but not the time step before, no ADC last time step
+        elif tb_tm1 and not tb_tm2:
+            prob_test = self.prob_test_tb
+        # presence of a non-TB WHO3 disease last time step, no ADC or TB last time step
+        elif non_tb_who3_tm1 and not tb_tm1:
+            prob_test = self.prob_test_non_tb_who3
+
+        return prob_test
 
-    def update_sub_pop_test_date(self, pop, sub_pop, prob_test):
+    def test_mark_non_hiv_symptomatic(self, pop):
         """
-        Update the last test date of a sub-population based on a given probability.
+        Mark non-HIV symptomatic individuals to undergo testing this time step.
         """
-        if len(sub_pop) > 0:
-            r = rng.uniform(size=len(sub_pop))
-            tested = r < prob_test
-            pop.set_present_variable(col.LAST_TEST_DATE, pop.date,
-                                     sub_pop=pop.apply_bool_mask(tested, sub_pop))
+        # testing occurs after a certain year if there is no covid disruption
+        if ((pop.date.year >= self.date_start_testing)
+           & (not (self.covid_disrup_affected | self.testing_disrup_covid))):
+
+            # undiagnosed (last time step) and untested population
+            not_diag_tested_pop = pop.get_sub_pop([(pop.get_correct_column(col.HIV_DIAGNOSED, dt=1), op.eq, False),
+                                                   (col.EVER_TESTED, op.eq, False),
+                                                   (col.TEST_MARK, op.eq, False)])
 
-    def update_anc_hiv_testing(self, pop, time_step):
+            if len(not_diag_tested_pop) > 0:
+                # mark people for testing
+                r = rng.uniform(size=len(not_diag_tested_pop))
+                s = rng.uniform(size=len(not_diag_tested_pop))
+                marked = ((r < (self.prob_test_non_tb_who3 + self.prob_test_who4)/2) &
+                          (s < self.prob_test_non_hiv_symptoms))
```

Comment:
I understand that `r` is related to the SAS line `u < (test_rate_non_tb_who3+test_rate_who4)/2 then do; `, but what does `s` /  `prob_test_non_hiv_symptoms` correspond to?